---
title: "Supplementary Figure 3. Joint Alignment and Visualization of Motifs from Fly, Mouse and Human HD Family"
output:
  pdf_document:
    fig_caption: yes
    toc: yes
    includes:
      in_header: header.tex
  html_document:
    fig_caption: yes
    self_contained: no
    toc: no
---

####Supplemental Figure 3. The joint comparison of motifs using MatAlign (3A) or MotIV (3B) between human (SELEX data, colored in brown), mouse (PBM data, colored in purple) and fly (B1H data, colored in green) HD domain proteins.The length of tree branches reflect motif distances, motifs within a distance threshold of 2.5 are clustered and a merged motif for that cluster is shown. Clusters are numbered by size, with the largest group listed as cluster 1. Branches leading to Drosophila motifs are highlighted in green, mouse motifs in brown and human motifs in purple. Motifs from the human and mouse often grouped by species despite the relatively small evolutionary distance between these organisms. This observation suggests that the motif generation method can have a substantial impact on motif clustering.


```{r setup, include=FALSE, echo=FALSE}
fn = local({
  function(x, letter) {
    paste('__Supplementary Figure 3', letter, '__: ', x, sep = '')
  }
})
```

```{r loadLib, message=FALSE, warning=FALSE, error=FALSE, echo=FALSE}
library(motifStack)
colorSet <- c("Dm"="#00FC00", "b1h"="#00FC00", "sw"="#008080", "bml"="darkgreen", 
              "Mm"="brown", "MmDREAM"="blue", "Ms"="#F69156", 
              "Hs"="#D900D9")

outpath <- "output"
matalign_path <- "./app/matalign-v4a"
neighbor_path <- "./app/neighbor.app/Contents/MacOS/neighbor"
MatAlign2tree_path <- "./MatAlign2tree.pl"

getMatAlignOut <- function(pcmpath, groupDistance=2.5, trim=0.2){
    pcms <- readPCM(pcmpath)
    pfms<-lapply(pcms,pcm2pfm)
    system(paste("perl", MatAlign2tree_path, "--in . --pcmpath", pcmpath, "--out", outpath,
             "--matalign", matalign_path, "--neighbor", neighbor_path, "--tree","UPGMA"))
    newickstrUPGMA <- readLines(con=file.path(outpath, "NJ.matalign.distMX.nwk"))
    phylog <- newick2phylog(newickstrUPGMA, FALSE)
    leaves <- names(phylog$leaves)
    motifs <- pfms[leaves]
    if(!is.na(groupDistance)){
        motifSig <- motifSignature(motifs, phylog, groupDistance=groupDistance,
                                   min.freq=1, trim=trim)
        sig <- signatures(motifSig)
        gpCol <- sigColor(motifSig)
    }else{
        motifSig <- NA
        sig <- NA
        gpCol <- NA
    }
    
    return(list(phylog=phylog, sig=sig, gpCol=gpCol, 
                motifs=DNAmotifAlignment(motifs), leaves=leaves, 
                unaligned.pfms=motifs))
}

pfmList2matrixList <- function(pfms){
        m <- lapply(pfms, function(.ele) as(.ele, "matrix"))
        names(m) <- unlist(lapply(pfms, function(.ele) .ele@name))
        m
}

getMotIVOut <- function(pcmpath, cc, align, groupDistance=.005, trim=0.2){
    pcms <- readPCM(pcmpath)
    pfms<-lapply(pcms,pcm2pfm)
    jaspar.scores <- 
        MotIV::readDBScores(
            file.path(".", "app", "scores", 
                      paste("JaspRand_", cc, "_", align, ".scores", sep="")))
    d <- MotIV::motifDistances(pfmList2matrixList(pfms), cc=cc, align=align)
    hc <- MotIV::motifHclust(d)
    phylog <- hclust2phylog(hc)
    pfms <- pfms[hc$order]
    aligned.pfms <- DNAmotifAlignment(pfms)
    leaveNames <-names(phylog$leaves)
    leaves <- names(phylog$leaves)
    motifs <- pfms[leaves]
    if(!is.na(groupDistance)){
        motifSig <- motifSignature(motifs, phylog, groupDistance=groupDistance,
                                   min.freq=1, trim=trim)
        sig <- signatures(motifSig)
        gpCol <- sigColor(motifSig)
    }else{
        motifSig <- NA
        sig <- NA
        gpCol <- NA
    }
    
    return(list(phylog=phylog, sig=sig, gpCol=gpCol, motifs=DNAmotifAlignment(motifs),
                leaves=leaves, unaligned.pfms=motifs))
}

```

```{r FlyMousePBMHuman, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Alignment of motifs from fly, mouse and human by MatAlign", "A"), echo=FALSE}
gpDis <- 2.5
# Supplementary Figure 3I. Align motifs from fly, mouse and human by MatAlign
FlyMousePBMHuman <- getMatAlignOut("pcmsDatasetFlyMousePBMHuman",
                                   groupDistance=gpDis)
attach(FlyMousePBMHuman)

##set methods color
species <- factor(gsub("^(Dm|Mm|Hs).*$", "\\1", leaves))
##c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", "Mm/Uniprobe/PBM/Seed-And-Wobble")
levels(species) <- colorSet[c("Dm", "Hs", "Mm")] 
species <- as.character(species)

leaveNames <- gsub("^(Dm|Mm|Hs)_", "", leaves)

plotMotifStackWithRadialPhylog(phylog=phylog, pfms=sig, 
                               col.bg=species, col.bg.alpha=.3, 
                               col.inner.label.circle=gpCol, 
                               inner.label.circle.width=0.02, 
                               labels.leaves=leaveNames, 
                               cleaves=.1, circle=1.5, circle.motif=1.8, 
                               clabel.leaves=.5, motifScale="logarithmic", 
                               angle=358, plotIndex=TRUE, IndexCex=.6, 
                               groupDistance=gpDis)

legend(-2, 2.2, 
       legend=c("Dm/FFS/B1H", 
                "Mm/Uniprobe/PBM/Seed-And-Wobble", 
                "Hs/CIS-BP/SELEX"), 
       fill= highlightCol(colorSet[c("Dm", "Hs", "Mm")], alpha=.3), 
       border="white", lty=NULL, bty = "n")

detach(FlyMousePBMHuman)
```

```{r FlyMousePBMHuman.MotIV, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Alignment of motifs from fly, mouse and human by MotIV", "B"), echo=FALSE}
gpDis <- 0.005

# Supplementary Figure 3J. Align motifs from fly vs mouse and human by MotIV
FlyMousePBMHuman.MotIV <- getMotIVOut("pcmsDatasetFlyMousePBMHuman", 
                                      cc="ALLR", align="SWU", 
                                      groupDistance=gpDis)
attach(FlyMousePBMHuman.MotIV)

##set methods color
species <- factor(gsub("^(Dm|Mm|Hs).*$", "\\1", leaves))
##c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", "Mm/Uniprobe/PBM/Seed-And-Wobble")
levels(species) <- colorSet[c("Dm", "Hs", "Mm")] 
species <- as.character(species)

leaveNames <- gsub("^(Dm|Mm|Hs)_", "", leaves)

plotMotifStackWithRadialPhylog(phylog=phylog, pfms=sig, 
                               col.bg=species, col.bg.alpha=.3, 
                               col.inner.label.circle=gpCol, 
                               inner.label.circle.width=0.02, 
                               labels.leaves=leaveNames, 
                               cleaves=.1, circle=1.5, circle.motif=1.8, 
                               clabel.leaves=.5, motifScale="logarithmic", 
                               angle=358, plotIndex=TRUE, IndexCex=.6, 
                               groupDistance=gpDis)

legend(-2, 2.2, 
       legend=c("Dm/FFS/B1H", 
                "Mm/Uniprobe/PBM/Seed-And-Wobble", 
                "Hs/CIS-BP/SELEX"), 
       fill= highlightCol(colorSet[c("Dm", "Hs", "Mm")], alpha=.3), 
       border="white", lty=NULL, bty = "n")

detach(FlyMousePBMHuman.MotIV)
```

```{r EN, fig.width=8, fig.height=4, dpi=72, fig.cap=fn("Alignment of engrailed orthologs by MatAlign", "C"), echo=FALSE}
gpDis <- 2

# Supplementary Figure 3G. Align motifs of engrailed orthologs by MatAlign
EN <- getMatAlignOut("pcmsDatasetEn", groupDistance=gpDis)
attach(EN)

##set methods color
species <- gsub("^(Dm|Hs|Mm).*?$", "\\1", leaves)
species[grepl("_M\\d+", leaves) & species=="Mm"] <- "MmDREAM"
species <- colorSet[species] ##c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", "Mm/Uniprobe/PBM/\nSeed-And-Wobble")

leaveNames <- gsub("^(Dm|Mm|Hs)_", "", leaves)

motifPiles(phylog=phylog, motifs, r.tree=.5, sig, 
           col.tree=species, col.leaves=species, 
           col.pfms2=gpCol, col.pfms2.width=.01, 
           labels.leaves=leaveNames, plotIndex=TRUE, 
           IndexCex=2, groupDistance=gpDis, clabel.leaves=2)
legend("topleft", legend=c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", 
                           "Mm/Uniprobe/PBM/\nSeed-And-Wobble", 
                           "Mm/CIS-BP/PBM/Dream5"), 
       fill=colorSet[c("Dm", "Hs", "Mm", "MmDREAM")], 
       border="white", lty=NULL, bty = "n", cex=.5)

detach(EN)
```


```{r EN.MotIV, fig.width=8, fig.height=4, dpi=72, fig.cap=fn("Alignment of engrailed orthologs by MotIV", "D"), echo=FALSE, eval=FALSE}
gpDis <- 0.005 

# Supplementary Figure 3H. Align motifs of engrailed orthologs by MotIV 
EN.MotIV <- getMotIVOut("pcmsDatasetEn", 
                                    cc="ALLR", align="SWU", 
                                    groupDistance=gpDis)
attach(EN.MotIV)

##set methods color
species <- gsub("^(Dm|Hs|Mm).*?$", "\\1", leaves)
species[grepl("_M\\d+", leaves) & species=="Mm"] <- "MmDREAM"
##c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", "Mm/Uniprobe/PBM/\nSeed-And-Wobble")
species <- colorSet[species]

leaveNames <- gsub("^(Dm|Mm|Hs)_", "", leaves)

motifPiles(phylog=phylog, motifs, r.tree=.5, sig, 
           col.tree=species, col.leaves=species, 
           col.pfms2=gpCol, col.pfms2.width=.01, 
           labels.leaves=leaveNames, plotIndex=TRUE, 
           IndexCex=2, groupDistance=gpDis, clabel.leaves=2)
legend(0.1, 0.65, legend=c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", 
                           "Mm/Uniprobe/PBM/\nSeed-And-Wobble", 
                           "Mm/CIS-BP/PBM/Dream5"), 
       fill=colorSet[c("Dm", "Hs", "Mm", "MmDREAM")], 
       border="white", lty=NULL, bty = "n", cex=.5)

detach(EN.MotIV)
```

```{r MmPBM_SELEX, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Motif signature and alignment of mouse data generated by different methods by MatAlign", "E"), echo=FALSE, eval=FALSE, eval=FALSE}
gpDis <- 2

# Supplementary Figure 3E. Align motifs from mouse generated by different methods (SELEX, PBM-SW and PBM-dream5) by MatAlign
MmPBM_SELEX <- getMatAlignOut("pcmsDatasetMmPBM_SELEX", groupDistance=gpDis)
attach(MmPBM_SELEX)

##set methods color
methods <- factor(grepl("_SELEX", leaves) + grepl("_M\\d{4}", leaves))
levels(methods) <- c("Mm", "MmDREAM", "Ms")
##c("PBM/Uniprobe/Seed-And-Wobble", "PBM/CIS-BP/DREAM5", "SELEX/CIS-BP")
levels(methods) <- colorSet[levels(methods)]
methods <- as.character(methods)

leaveNames <- gsub("_(SELEX|PBM)", "", gsub("_(Jolma|Berger)", "", leaves))

## calculate average of top 8 position information content for each motif
icgp <- sapply(sapply(unaligned.pfms, getIC), function(.ele) 
  mean(sort(.ele, decreasing=TRUE)[1:min(8, length(.ele))]))
icgp.ranges <- range(icgp)
icgp <- cut(icgp, 10, labels=colorRampPalette(c("green", "black", "red"))(10))
icgp.image <- as.raster(matrix(rev(levels(icgp)), ncol=1))
icgp <- as.character(icgp)

plotMotifStackWithRadialPhylog(phylog=phylog, pfms=sig, 
                               col.bg=methods, col.bg.alpha=.3, 
                               col.inner.label.circle=gpCol, 
                               inner.label.circle.width=0.02, 
                               labels.leaves=leaveNames, cleaves=.1, 
                               circle=.95, circle.motif=1.8, clabel.leaves=.8,
                               motifScale="logarithmic", angle=358, 
                               plotIndex=TRUE, IndexCex=.6, groupDistance=gpDis)

legend(-2, 2.2, legend=c("Mm/Uniprobe/PBM/Seed-And-Wobble", 
                         "Mm/CIS-BP/PBM/DREAM5", "Mm/CIS-BP/SELEX"), 
       fill=highlightCol(colorSet[c("Mm", "MmDREAM", "Ms")], .3), 
       border="white", lty=NULL, bty = "n")
```