---
title: "Supplementary Figure 2. Example Code Snippets for Visualizing Motif Alignments with motifStack "
output:
  pdf_document:
    fig_caption: yes
    toc: yes
    includes:
      in_header: header.tex
  html_document:
    fig_caption: yes
    self_contained: no
    toc: yes
---
```{r setup, include=FALSE}
fn = local({
  function(x, letter, auto=FALSE) {
    paste('__Supplementary Figure 2', letter, '__: ', x, sep = '')
  }
})
```


```{r inputdata, message=FALSE, warning=FALSE, error=FALSE}
##load the library
library(motifStack)

##read pcms (position count matrices)
pcmpath <- "pcmsDatasetDM"
pcms <- readPCM(pcmpath)
## color blindness
source("safeColor.R")
pcms <- safeColor(pcms)
##convert to pfms (position frequency matrices)
pfms<-lapply(pcms,pcm2pfm)
```

## Supplementary Figure 2A-D. Examples of linear and radial dendrograms and code to generate them.

### Plot motifs as a stack.
```{r stackLayout, fig.width=6, fig.height=6, dpi=72, fig.cap=fn("Plotted as a stack of motifs", "A")}
subset_pfms <- pfms[10:15]
motifStack(subset_pfms, layout="stack")
```

\clearpage

###Plot motifs as a linear tree with nucleotides drawn proportional to their frequency.

```{r phylogLayout, fig.width=6, fig.height=6, dpi=72, fig.cap=fn("Plotted as a linear tree with nucleotides drawn proportional to their frequency", "B")}
##try different style of sequence logo
motifStack(pfms[sample(1:length(pfms), 20)], layout="phylog", clabel.leaves=.8,
           f.logo=0.5, ic.scale=FALSE)
```

\clearpage

###Plot motifs as a linear tree depicting information content (IC).

```{r treeLayout, fig.width=6, fig.height=6, dpi=72, fig.cap=fn("Plotted as a linear tree with IC", "C")}
##By default, MotIV is used for clustering
motifStack(subset_pfms, layout="tree", trueDist=TRUE)
```

\clearpage

### Plot motifs as a radial tree.

```{r radialPhylogLayout, fig.width=10, fig.height=10, dpi=72, fig.cap=fn("Plotted as a radial tree", "D")}
motifStack(pfms, layout="radialPhylog")
```

\clearpage

## Supplementary Figure 2E-G. Merge motifs with different distance cutoffs and display these motif signatures in various layouts.

Before motifs are merged, the distances of the motifs are calculated using STAMP, MovIV or MatAlign. Here we show different ways to display the merged motifs and how different distance cutoffs affect the resulting motif sigatures. The red dotted line indicates the distance cutoff used.

### Merge motifs using different distance cutoffs.

```{r merging, fig.width=4, fig.height=4, fig.cap=fn("Merge motifs by different distance", "E")}
outpath <- "output"
matalign_path <- "./app/matalign-v4a"
neighbor_path <- "./app/neighbor.app/Contents/MacOS/neighbor"
MatAlign2tree_path <- "./MatAlign2tree.pl"
pcmpath <- "pcmsDatasetDM"
pcms <- readPCM(pcmpath)
pcms <- safeColor(pcms)
pfms<-lapply(pcms,pcm2pfm)
system(paste("perl", MatAlign2tree_path, "--in . --pcmpath", pcmpath, 
             "--out", outpath,
             "--matalign", matalign_path, 
             "--neighbor", neighbor_path, 
             "--tree","UPGMA"))
newickstrUPGMA <- readLines(con=file.path(outpath, "NJ.matalign.distMX.nwk"))
phylogUPGMAmatAlign <- newick2phylog(newickstrUPGMA, FALSE)
##get the leaves of tree for ordering the pfms
matAlignLeaveNames <- names(phylogUPGMAmatAlign$leaves)
this_motifs <- pfms[matAlignLeaveNames]
matAlignLeaveNames <- gsub("^Dm_", "", matAlignLeaveNames)

for(groupDistance in c(0, 1, 2, 3)){
    motifSig <- motifSignature(this_motifs, 
                               phylogUPGMAmatAlign, 
                               groupDistance=groupDistance, 
                               min.freq=1)
    this_sig <- signatures(motifSig)
    ## get color set for the signature groups
    this_gpCol <- sigColor(motifSig)
    plotMotifStackWithRadialPhylog(phylog=phylogUPGMAmatAlign, 
                                   pfms=this_sig,
                                   col.inner.label.circle=this_gpCol, 
                                   inner.label.circle.width=0.02, 
                                   labels.leaves=matAlignLeaveNames, 
                                   cleaves=.2, circle=1, circle.motif=1.5, 
                                   clabel.leaves=.4, motifScale="logarithmic", 
                                   angle=358, plotIndex=TRUE, IndexCex=.6, 
                                   groupDistance=groupDistance,
                                   groupDistanceLineCol="#D55E00")
    text(0, 2.3, label=paste("group distance =", groupDistance), cex=2)
}
```

```{r fig.cap=fn("Motif signature plotted as a radial tree by different distance", "E"), fig.width=0.1, fig.height=0.02, echo=FALSE}
par(mar=rep(0, 4))
plot.new()
```

\clearpage

### Plot motif signature as a word cloud in circle.
```{r motifCloud, warning=FALSE, fig.width=7.5, fig.height=7.5, dpi=72, fig.cap=fn("Motif signature plotted as a word cloud with motif size representing the number of motifs contributed to the signature. The larger the size, the larger the number of motifs that share the motif signature", "F")}
## get signature
gpDist <- c(.5, 1)
motifSig <- lapply(gpDist, motifSignature, 
                   pfms=this_motifs, 
                   phylog=phylogUPGMAmatAlign, 
                   min.freq=1)
## motif cloud, cloud style
motifCloud(motifSig[[1]], layout="cloud", scale=c(9, .75))
```

\clearpage

###Plot motif signature as a word cloud in rectangle.
```{r motifCloudRect, warning=FALSE, fig.width=3.5, fig.height=3.5, fig.cap=fn("Motif signature plotted as a word cloud in rectangular by different distances", "G")}
## motif cloud, rectangle style
for(i in 1:2){
    motifCloud(motifSig[[i]], layout="rectangles", ic.scale=FALSE)
    op <- par(mar = c(0, 0, 0, 0))
    text(.5, .985, label=paste("group distance =", gpDist[i]))
    par(op)
}
```

```{r fig.cap=fn("Motif signature plotted as a word cloud in rectangular by different distances", "G"), fig.width=0.1, fig.height=0.02, echo=FALSE}
par(mar=rep(0, 4))
plot.new()
```

\clearpage

## Supplementary Figure 2H-M. Use various color options to highlight different motif features.

motifStack offers five options to color a radial tree: the background of the inner circle, the text of motif names, the background of the motif names, the inner label ring and the outer label ring. Here are a few examples using these options to show the motif's data source, the computational algorithm that generated the motifs, the motif's information content (IC) and the TF. 

```{r functions}
getPairColor <- function(n=10L){
    if(n %% 2 != 0) n <- n+1
    n <- n/2
    n <- rainbow(n)#as.character(t(matrix(rainbow(n=n), ncol=2, byrow=FALSE)))
    n2 <- highlightCol(n, .5)
    as.character(t(cbind(n, n2)))
}

pairColor <- getPairColor(22)

pfmList2matrixList <- function(pfms){
        m <- lapply(pfms, function(.ele) as(.ele, "matrix"))
        names(m) <- unlist(lapply(pfms, function(.ele) .ele@name))
        m
}
getMotIVOut <- function(pfms, cc, align){
    jaspar.scores <- 
        MotIV::readDBScores(
            file.path(".", "app", "scores", 
                      paste("JaspRand_", cc, "_", align, ".scores", sep="")))
    d <- MotIV::motifDistances(pfmList2matrixList(pfms), cc=cc, align=align)
    hc <- MotIV::motifHclust(d, method="average")
    phylog <- hclust2phylog(hc)
    pfms <- pfms[hc$order]
    aligned.pfms <- DNAmotifAlignment(pfms)
    leaveNames <-names(phylog$leaves)
    ## data source
    dataSource <- factor(grepl("_M", leaveNames))
    levels(dataSource) <- c("#F0E442", "#56B4E9") ##c("Uniprobe", "CIS-BP")
    dataSource <- as.character(dataSource)
    ## algorithm
    algorithm <- 
        factor((!grepl("_M", leaveNames)) + grepl("_bml", leaveNames))
    levels(algorithm) <- 
        c("#0072B2", "#CC79A7", "#E69F00") ##("DREAM5", "Seed-And-Wobble", "BEEML")
    algorithm <- as.character(algorithm)
    ## motifs from same PBM data
    motifGroup <- factor(gsub("(.*?)_.*$", "\\1", leaveNames))
    levels.motifGroup <- levels(motifGroup)
    levels(motifGroup) <- pairColor[1:length(levels(motifGroup))]
    colors.motifGroup <- levels(motifGroup)
    motifGroup <- as.character(motifGroup)
    return(list(aligned.pfms=aligned.pfms, unaligned.pfms=pfms, phylog=phylog, 
                leaveNames=leaveNames, 
                dataSource=dataSource,
                algorithm=algorithm,
                motifGroup=motifGroup,
                levels.motifGroup=levels.motifGroup,
                colors.motifGroup=colors.motifGroup))
}
##read pcms
pcmpath <- "pcmsDatasetAlgorithm"
pcms <- readPCM(pcmpath)
pcms <- safeColor(pcms)
##convert to pfms
pfms<-lapply(pcms,pcm2pfm)
```

###Use the color of the background of inner circle to distinguish different data sources with the parameter col.bg.
```{r colorsets1, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Use the color of the background of inner circle to distinguish different data sources with the parameter col.bg", "H")}
motIVout <- getMotIVOut(pfms, "PCC", "SWU")
attach(motIVout)
plotMotifStackWithRadialPhylog(phylog=phylog, pfms=unaligned.pfms,
                            labels.leaves=leaveNames, 
                            col.bg=dataSource, col.bg.alpha=.3,
                            cleaves=.2, circle=1.1, circle.motif=1.6, 
                            clabel.leaves=.8, angle=358)
legend(-2.3, 2.4, legend=c("Uniprobe", "CIS-BP"), 
       fill= highlightCol(c("#F0E442", "#56B4E9"), alpha=.3), 
       border="white", lty=NULL, bty = "n", cex=1)

```


\clearpage

###Use the color of text of motif names to distinguish the computational algorithms used with the parameter col.leaves.
```{r colorsets2, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Use the color of the text of motif names to separate the computational algorithms used with the parameter col.leaves", "I")}
plotMotifStackWithRadialPhylog(phylog=phylog, pfms=unaligned.pfms,
                            labels.leaves=leaveNames, 
                            col.leaves=algorithm,
                            cleaves=.2, circle=1.1, circle.motif=1.6, 
                            clabel.leaves=.8, angle=358)
legend(-2.3, 2.4, legend=c("DREAM5", "Seed-And-Wobble", "BEEML"), 
       fill= c("#0072B2", "#CC79A7", "#E69F00"), 
       border="white", lty=NULL, bty = "n", cex=1)
```

\clearpage

###Use the color of the background of the motif names to distinguish motifs with high and low IC with the parameter col.leaves.bg.
```{r colorsets3, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Use the color of the background of the motif names to distinguish motifs with high and low IC with the parameter col.leaves.bg", "J")}
icgp <- ifelse(sapply(sapply(unaligned.pfms, getIC), mean) > 0.6, 
               "lightgray", "black")
plotMotifStackWithRadialPhylog(phylog=phylog, pfms=unaligned.pfms,
                            labels.leaves=leaveNames, 
                            col.leaves.bg=icgp, 
                            col.leaves.bg.alpha=.3, 
                            cleaves=.2, circle=1.1, circle.motif=1.6, 
                            clabel.leaves=.8, angle=358)
legend(-2.3, 2.4, 
       legend=c("Information Content <= 0.6", "Information Content > 0.6"), 
       fill= highlightCol(c("black", "lightgray"), alpha=.3), 
       border="white", lty=NULL, bty = "n", cex=1)

```

\clearpage

###Use the color of the inner ring of the motif circle to distinguish the TF with the parameter col.inner.label.circle.
```{r colorsets4, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Use the color of the inner ring of the motif circle to distinguish the TFs with the parameter col.inner.label.circle", "K")}
plotMotifStackWithRadialPhylog(phylog=phylog, pfms=unaligned.pfms,
                            labels.leaves=leaveNames, 
                            col.inner.label.circle=motifGroup,
                            inner.label.circle.width=0.1,
                            cleaves=.2, circle=1.1, circle.motif=1.6, 
                            clabel.leaves=.8, angle=358)
legend(1.5, 2.4, legend=levels.motifGroup[1:12], 
       fill= colors.motifGroup[1:12], 
       border="black", lty=NULL, bty = "n", ncol=2, cex=.8)
legend(1.5, -2, legend=levels.motifGroup[13:21], 
       fill= colors.motifGroup[13:21], 
       border="black", lty=NULL, bty = "n", ncol=2, cex=.8)
```

\clearpage

###Use the color of the outer ring of the motif circle to distinguish the TFs with the parameter col.outer.label.circle.
```{r colorsets5, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Use the color of the outer ring of the motif circle to distinguish the TFs with parameter col.outer.label.circle", "L")}
plotMotifStackWithRadialPhylog(phylog=phylog, pfms=unaligned.pfms,
                            labels.leaves=leaveNames, 
                            col.outer.label.circle=motifGroup,
                            outer.label.circle.width=0.1,
                            cleaves=.2, circle=1.1, circle.motif=1.6, 
                            clabel.leaves=.8, angle=358)
legend(1.5, 2.4, legend=levels.motifGroup[1:12], 
       fill= colors.motifGroup[1:12], 
       border="black", lty=NULL, bty = "n", ncol=2, cex=.8)
legend(1.5, -2, legend=levels.motifGroup[13:21], 
       fill= colors.motifGroup[13:21], 
       border="black", lty=NULL, bty = "n", ncol=2, cex=.8)

```

\clearpage

###Use a combination of coloring options to visualize multiple motif features.
```{r colorsets6, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Use a combination of coloring options to visualize multiple motif feature", "M")}
plotMotifStackWithRadialPhylog(phylog=phylog, pfms=aligned.pfms,
                               labels.leaves=leaveNames, 
                               col.bg=algorithm, col.bg.alpha=.3,
                               col.leaves.bg=icgp, 
                               col.leaves.bg.alpha=.3, 
                               cleaves=.2, circle=1.1, circle.motif=1.6, 
                               clabel.leaves=.8, angle=358)
legend(-2.3, 2.4, 
       legend=c("algorithm", "Mm/PBM/DREAM5", "Mm/PBM/Seed-And-Wobble", 
                "Mm/PBM/BEEML", "Information Content", "<= 0.6", "> 0.6"), 
       fill= c("white", 
               highlightCol(c("#0072B2", "#CC79A7", "#E69F00"), alpha=.3), 
               "white", highlightCol(c("black", "lightgray"), alpha=.3)), 
       border="white", lty=NULL, bty = "n", cex=.8)

detach(motIVout)
```

\clearpage

## Supplementary Figure 2N-O. Compare different Column Comparison Metrics (CCM) and aligment methods.

Here we used a subset of mouse HD TFs with PBM data. For each TF, we generated three different motifs using three different computational methods. We compared how well different CCM and alignment methods could group together the motifs from the same TFs. This example illustrates how to set the CCM and the alignment method, using MotIV or MatAlign, and that different methods can substantially affect the resulting alignment as visulaized in motifStack.

When clustering using MovIV, Pearson Correlation Coefficient (PCC) or
Average Log Likelihood Ratio (ALLR) was used as the CCM, and Smith-Waterman Ungapped (SWU) or Needleman-Wunsch (NW) was used as the alignment method. When clustering the motifs using MatAlign, ALLR and SWU were used as CCM and alignment respectively.

To run the MatAlign examples, phylip (http://evolution.genetics.washington.edu/phylip/progs.data.dist.html) and MatAlign (http://stormo.wustl.edu/MatAlign/) need to be installed first.

###Visualize motif alignments generated by MotIV with different CCM and alignment methods.
```{r MotIV, fig.width=3.5, fig.height=3.5, fig.cap=fn("Alignment of mouse PBM data by MotIV with different CCM and alignment methods", "N")}

sta <- mapply(function(cc, align){
    motIVout <- getMotIVOut(pfms, cc, align)
    attach(motIVout)
    plotMotifStackWithRadialPhylog(phylog=phylog, pfms=aligned.pfms,
                               labels.leaves=leaveNames, 
                               col.bg=algorithm, col.bg.alpha=.3, 
                               col.inner.label.circle=motifGroup,
                               inner.label.circle.width=0.1,
                               cleaves=.2, circle=1.1, circle.motif=1.6, 
                               clabel.leaves=.3, angle=358)

    legend(-2.3, 2.4, 
           legend=c("Mm/PBM/DREAM5", "Mm/PBM/Seed-And-Wobble", 
                    "Mm/PBM/BEEML"), 
           fill= highlightCol(c("#0072B2", "#CC79A7", "#E69F00"), alpha=.3), 
           border="white", lty=NULL, bty = "n", cex=.5)
    text(0, 2.3, label=paste("cc=", cc, "; align=", align), cex=1.5)
    cnt <- rle(motifGroup)
    cnt <- split(algorithm, rep(1:length(cnt$lengths), cnt$lengths))
    cnt <- table(sapply(cnt, function(.ele) length(unique(.ele))))
    cnt.1 <- vector("integer", 3)
    names(cnt.1) <- 1:3
    cnt.1[names(cnt)] <- cnt
    cnt.1["1"] <- (cnt.1["1"]+cnt.1["2"]*2+cnt.1["3"]*3)/3 - cnt.1["2"] - cnt.1["3"]
    text(0, -2.3, 
         label=paste(names(cnt.1), cnt.1, sep=":", collapse="; "), cex=1.5)
    detach(motIVout)
}, c("PCC", "PCC", "ALLR", "ALLR"), c("NW", "SWU", "NW", "SWU"))
```

```{r fig.cap=fn("Alignment of mouse PBM data by MotIV with different CCM and alignment methods", "N"), fig.width=0.1, fig.height=0.02, echo=FALSE}
par(mar=rep(0, 4))
plot.new()
```

\clearpage

The number (1:X; 2:Y; 3:Z) on the bottom of the figure 2O and 2P represents the number of motifs from the same TF grouped by different algorithms. 3:Z means that Z number of TFs with 3 of 3 motifs clustered together; 2:Y means that Y number of TFs with 2 of 3 motifs clustered together; 1:X means that X number of TFs failed to cluster any of the 3 motifs together. Thus, the lower the X, the higher the Y and Z, the better the method performed. In this example, MatAlign performed better than MotIV even using the same CCM (ALLR) and alignment method (SWU). 

###Visualize motif alignments generated by MatAlign with ALLR as the MCC and SWU as the alignment method.
```{r MatAlign, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Alignment of mouse PBM data by MatAlign", "O")}
##read newick tree. Alignment is done by MatAlign
##The newick tree is generated by Neighbor, which is a part of phylip
outpath <- "output"
matalign_path <- "./app/matalign-v4a"
neighbor_path <- "./app/neighbor.app/Contents/MacOS/neighbor"
MatAlign2tree_path <- "./MatAlign2tree.pl"
system(paste("perl", MatAlign2tree_path, "--in . --pcmpath", pcmpath, 
             "--out", outpath,
             "--matalign", matalign_path, 
             "--neighbor", neighbor_path, 
             "--tree","UPGMA"))
newickstrUPGMA <- readLines(con=file.path(outpath, "NJ.matalign.distMX.nwk"))
##convert it to phylog object
phylogUPGMAmatAlign <- newick2phylog(newickstrUPGMA, FALSE)

##get the leaves of phylog for reordering the pfms
leaveNames <- names(phylogUPGMAmatAlign$leaves)
this_motifs <- pfms[leaveNames]

## data source
dataSource <- factor(grepl("_M", leaveNames))
levels(dataSource) <- c("#F0E442", "#56B4E9") ##c("Uniprobe", "CIS-BP")
dataSource <- as.character(dataSource)
## algorithm
algorithm <- factor((!grepl("_M", leaveNames)) + grepl("_bml", leaveNames))
levels(algorithm) <- c("#0072B2", "#CC79A7", "#E69F00") ##("DREAM5", "Seed-And-Wobble", "BEEML")
algorithm <- as.character(algorithm)
## motifs from the same PBM data source
motifGroup <- factor(gsub("(.*?)_.*$", "\\1", leaveNames))
levels.motifGroup <- levels(motifGroup)
levels(motifGroup) <- pairColor[1:length(levels(motifGroup))]
colors.motifGroup <- levels(motifGroup)
motifGroup <- as.character(motifGroup)

## draw the motifs
plotMotifStackWithRadialPhylog(phylog=phylogUPGMAmatAlign, 
                               pfms=DNAmotifAlignment(this_motifs),
                               labels.leaves=leaveNames, 
                               col.bg=algorithm, 
                               col.bg.alpha=.3, 
                               col.inner.label.circle=motifGroup, 
                               inner.label.circle.width=.1,
                               cleaves=.2, circle=1.1, 
                               circle.motif=1.6, 
                               clabel.leaves=.6, angle=358)

legend(1.5, 2.4, legend=levels.motifGroup[1:12], 
       fill= colors.motifGroup[1:12], 
       border="black", lty=NULL, bty = "n", ncol=2, cex=.8)
legend(1.5, -2, legend=levels.motifGroup[13:21], 
       fill= colors.motifGroup[13:21], 
       border="black", lty=NULL, bty = "n", ncol=2, cex=.8)
legend(-2.35, 2.4, 
       legend=c("Mm/PBM/DREAM5", "Mm/PBM/Seed-And-Wobble", "Mm/PBM/BEEML"),
       fill= highlightCol(c("#0072B2", "#CC79A7", "#E69F00"), alpha=.3), 
       border="white", lty=NULL, bty = "n", cex=.8)
cnt <- rle(motifGroup)
cnt <- split(algorithm, rep(1:length(cnt$lengths), cnt$lengths))
cnt <- table(sapply(cnt, function(.ele) length(unique(.ele))))
cnt.1 <- vector("integer", 3)
names(cnt.1) <- 1:3
cnt.1[names(cnt)] <- cnt
cnt.1["1"] <- (cnt.1["1"]+cnt.1["2"]*2+cnt.1["3"]*3)/3 - cnt.1["2"] - cnt.1["3"]
text(0, -2.3, label=paste(names(cnt.1), cnt.1, sep=":", collapse="; "), cex=1.5)
```

\clearpage

## Supplementary Figure 2P. Improved alignment of fly HD family motifs using MatAlign compared to MotIV. The MatAlign alignment method is superior for discriminating between the closely related motifs in the fly HD family.

```{r revolution.1, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("MotIV(ALLR and SWU) as the alignment method", "P.a")}
pcmpath <- "pcmsDatasetDM"
pcms <- readPCM(pcmpath)
pcms <- safeColor(pcms)
pfms<-lapply(pcms,pcm2pfm)
motIVout <- getMotIVOut(pfms, "ALLR", "SWU")
plotMotifStackWithRadialPhylog(phylog=motIVout$phylog, 
                               pfms=motIVout$aligned.pfms, 
                               labels.leaves=motIVout$leaveNames, 
                               cleaves=.2, circle=1.2, circle.motif=1.6, 
                               clabel.leaves=1, 
                               motifScale="logarithmic", 
                               angle=358, 
                               plotIndex=FALSE)
text(0, 2.4, label="motIV: cc=ALLR; align=SWU", cex=1.5)
```

\clearpage

```{r revolution.2, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("MatAlign (ALLR and SWU) as the alignment method", "P.b")}
## function to read example data
getMatAlignOut <- function(pcmpath, groupDistance=NA, trim=0.2){
    pcms <- readPCM(pcmpath)
    pcms <- safeColor(pcms)
    pfms<-lapply(pcms,pcm2pfm)
    source("zzz.R")
    system(paste("perl", MatAlign2tree_path, "--in . --pcmpath", pcmpath, 
                 "--out", outpath,
                 "--matalign", matalign_path, 
                 "--neighbor", neighbor_path, 
                 "--tree","UPGMA"))
    newickstrUPGMA <- 
        readLines(con=file.path(outpath, "NJ.matalign.distMX.nwk"))
    phylog <- newick2phylog(newickstrUPGMA, FALSE)
    leaves <- names(phylog$leaves)
    motifs <- pfms[leaves]
    if(!is.na(groupDistance)){
        motifSig <- 
            motifSignature(motifs, phylog, 
                           groupDistance=groupDistance, 
                           min.freq=1, trim=trim)
        sig <- signatures(motifSig)
        gpCol <- sigColor(motifSig)
    }else{
        motifSig <- NA
        sig <- NA
        gpCol <- NA
    }
    
    return(list(phylog=phylog, sig=sig, gpCol=gpCol, 
                motifs=DNAmotifAlignment(motifs), 
                leaves=leaves, 
                unaligned.pfms=motifs))
}

matAlignOut <- getMatAlignOut(pcmpath)
plotMotifStackWithRadialPhylog(phylog=matAlignOut$phylog, 
                               pfms=matAlignOut$motifs, 
                               labels.leaves=matAlignOut$leaves, 
                               cleaves=.2, circle=1.2, circle.motif=1.6, 
                               clabel.leaves=1, motifScale="logarithmic", 
                               angle=358, plotIndex=FALSE)
text(0, 2.4, label="MatAlign", cex=1.5)
```

\clearpage

## Supplementary Figure 2Q. Import PFMs/PCMs files in Transfac, CisBP, or JASPAR format in batch mode.

```{r importMatrix, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Aligned motifs imported from PFMs/PCMs files in Transfac, CisBP, or JASPAR format in batch mode.", "Q"), message=FALSE, warning=FALSE, error=FALSE}
path <- system.file("extdata", package = "motifStack", mustWork = TRUE)
pcms <- importMatrix(dir(path, "*.pcm", full.names = TRUE), format = "pcm", to = "pcm")
JASPAR <- importMatrix(dir(path, "*.jaspar", full.names = TRUE))
pfms <- importMatrix(dir(path, "*.pfm", full.names = TRUE))
transfac <- importMatrix(file.path(path, c("transfac.like.test.transfac", "RUNX1.transfac")))
cisbp <- importMatrix(file.path(path, "PWM.cisbp"))
motifs <- unlist(c(pcms, JASPAR, pfms, transfac, cisbp))
motifs[sapply(motifs, class)=="pcm"] <- 
  lapply(motifs[sapply(motifs, class)=="pcm"], pcm2pfm)
motifs <- safeColor(motifs)
motifStack(motifs, layout = "phylog")
```

\clearpage

## Supplementary Figure 2R. Plot affinity logo from a position specific affinity matrix (PSAM) as described by Foat et al. (Barrett C. Foat, Alexandre V. Morozov, Harmen J. Bussemaker; Statistical mechanical modeling of genome-wide transcription factor occupancy data by MatrixREDUCE, Bioinformatics, Volume 22, Issue 14, 15 July 2006, Pages e141-e149, https://doi.org/10.1093/bioinformatics/btl223).

```{r plotAffinityLogo, fig.width=12, fig.height=4, dpi=72, fig.cap=fn("PSAM plotted as an affinity logo.", "R"), message=FALSE, warning=FALSE, error=FALSE}
psam <- importMatrix(file.path(path, "PSAM.mxr"), format = "psam")[[1]]
psam <- safeColor(psam)
motifStack(psam)
```

\clearpage

##Session information including the version of R , motifStack and other packages.

```{r}
sessionInfo()
```
