---
title: "Supplementary Figure 3. Alignment and Visualization of Motifs from Fly, Mouse and Human HD Family"
output:
  pdf_document:
    fig_caption: yes
    toc: yes
    includes:
      in_header: header.tex
  html_document:
    fig_caption: yes
    self_contained: no
    toc: no
---

__Supplemental Figure 3A__: A linear dendrogram of all Drosophila HD motifs illustrates how motifStack can be used to visualize the relationships between TF motifs, including aligned motifs for individual TFs (center) and motif signatures for motif clusters (far right), where the logo size scales with the number of members within the cluster. Motif clusters are determined using a distance threshold of 2 (dashed magenta line), and are delineated using alternating light and dark grey bars that are numbered based on cluster size.

__Supplemental Figure 3B__: Alignment of motifs from fly, mouse and human by MatAlign. Joint comparison of motifs with MatAlign using motifs from human (SELEX data, colored in brown), mouse (PBM data, colored in purple) and fly (B1H data, colored in turquoise) HD domain proteins. The length of tree branches reflect motif distances. Motif clusters determined with a distance threshold of 2.5 are indicated by the alternating light and dark grey bars that are numbered by size, with the largest group listed as cluster 1. A single merged motif is shown for each cluster.

__Supplemental Figure 3C__: Alignment of motifs from fly, mouse and human by MotIV. Joint comparison of motifs with MotIV using a distance threshold of 0.005 with the same data sets and display settings as Supplemental Figure 3B.

__Supplemental Figure 3D__: HD motifs of Engrailed homologs from Drosophila, mouse and human illustrates how different experimental and computational methods can influence the resulting binding site motifs for orthologous proteins. Motif clusters are determined using a distance threshold of 2 (dashed magenta line). Dm/FFS/B1H (colored in turquoise) represents fly B1H binding site data. Hs/CIS-BP/SELEX (colored in purple) represents human SELEX data. Mm/Uniprobe/PBM/Seed-And-Wobble (colored in magenta) represents mouse PBM data analyzed by the Seed and Wobble motif discovery tool. Mm/CIS-BP/PBM/Dream5 (colored in blue) is mouse PBM data analyzed by the Dream5 motif discovery tool.


```{r setup, include=FALSE, echo=FALSE}
fn = local({
  function(x, letter) {
    paste('__Supplementary Figure 3', letter, '__: ', x, sep = '')
  }
})
library(knitr)
opts_chunk$set(warning=FALSE, message=FALSE)
```

```{r loadLib, message=FALSE, warning=FALSE, error=FALSE, echo=FALSE}
library(motifStack)
source("safeColor.R")

outpath <- "output"
matalign_path <- "./app/matalign-v4a"
neighbor_path <- "./app/neighbor.app/Contents/MacOS/neighbor"
MatAlign2tree_path <- "./MatAlign2tree.pl"

getMatAlignOut <- function(pcmpath, groupDistance=2.5, trim=0.2){
    pcms <- readPCM(pcmpath)
    pcms <- safeColor(pcms)
    pfms<-lapply(pcms,pcm2pfm)
    system(paste("perl", MatAlign2tree_path, "--in . --pcmpath", pcmpath, "--out", outpath,
             "--matalign", matalign_path, "--neighbor", neighbor_path, "--tree","UPGMA"))
    newickstrUPGMA <- readLines(con=file.path(outpath, "NJ.matalign.distMX.nwk"))
    phylog <- newick2phylog(newickstrUPGMA, FALSE)
    leaves <- names(phylog$leaves)
    motifs <- pfms[leaves]
    if(!is.na(groupDistance)){
        motifSig <- motifSignature(motifs, phylog, groupDistance=groupDistance,
                                   min.freq=1, trim=trim)
        sig <- signatures(motifSig)
        gpCol <- sigColor(motifSig)
    }else{
        motifSig <- NA
        sig <- NA
        gpCol <- NA
    }
    
    return(list(phylog=phylog, sig=sig, gpCol=gpCol, 
                motifs=DNAmotifAlignment(motifs), leaves=leaves, 
                unaligned.pfms=motifs))
}

pfmList2matrixList <- function(pfms){
        m <- lapply(pfms, function(.ele) as(.ele, "matrix"))
        names(m) <- unlist(lapply(pfms, function(.ele) .ele@name))
        m
}

getMotIVOut <- function(pcmpath, cc, align, groupDistance=.005, trim=0.2){
    pcms <- readPCM(pcmpath)
    pcms <- safeColor(pcms)
    pfms<-lapply(pcms,pcm2pfm)
    jaspar.scores <- 
        MotIV::readDBScores(
            file.path(".", "app", "scores", 
                      paste("JaspRand_", cc, "_", align, ".scores", sep="")))
    d <- MotIV::motifDistances(pfmList2matrixList(pfms), cc=cc, align=align)
    hc <- MotIV::motifHclust(d)
    phylog <- hclust2phylog(hc)
    pfms <- pfms[hc$order]
    aligned.pfms <- DNAmotifAlignment(pfms)
    leaveNames <-names(phylog$leaves)
    leaves <- names(phylog$leaves)
    motifs <- pfms[leaves]
    if(!is.na(groupDistance)){
        motifSig <- motifSignature(motifs, phylog, groupDistance=groupDistance,
                                   min.freq=1, trim=trim)
        sig <- signatures(motifSig)
        gpCol <- sigColor(motifSig)
    }else{
        motifSig <- NA
        sig <- NA
        gpCol <- NA
    }
    
    return(list(phylog=phylog, sig=sig, gpCol=gpCol, motifs=DNAmotifAlignment(motifs),
                leaves=leaves, unaligned.pfms=motifs))
}

```


```{r Fly_DNA, fig.width=8, fig.height=11, dpi=72, fig.cap=fn("A linear dendrogram of all Drosophila HD motifs illustrates how motifStack can be used to visualize the relationships between TF motifs, including aligned motifs for individual TFs (center) and motif signatures for motif clusters (far right), where the logo size scales with the number of members within the cluster. Motif clusters are determined using a distance threshold of 2 (dashed magenta line), and are delineated using alternating light and dark grey bars that are numbered based on cluster size.", "A"), echo=FALSE}
gpDis <- 2
Fly_DNA <- getMatAlignOut("pcmsDatasetFly", groupDistance=gpDis)
attach(Fly_DNA)

leaveNames <- gsub("^Dm_", "", leaves)

motifPiles(phylog=phylog, motifs, sig, 
           col.pfms=gpCol, col.pfms.width=.01,
           col.pfms2=gpCol, col.pfms2.width=.01, 
           labels.leaves=leaveNames, 
           plotIndex=c(FALSE, TRUE), IndexCex=1, 
           groupDistance=gpDis, clabel.leaves=1,
           groupDistanceLineCol='#D55E00')

detach(Fly_DNA)
```


```{r FlyMousePBMHuman, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Alignment of motifs from fly, mouse and human by MatAlign. Joint comparison of motifs with MatAlign using motifs from human (SELEX data, colored in brown), mouse (PBM data, colored in purple) and fly (B1H data, colored in turquoise) HD domain proteins. The length of tree branches reflect motif distances. Motif clusters determined with a distance threshold of 2.5 are indicated by the alternating light and dark grey bars that are numbered by size, with the largest group listed as cluster 1. A single merged motif is shown for each cluster.", "B"), echo=FALSE}
gpDis <- 2.5
# Supplementary Figure 3I. Align motifs from fly, mouse and human by MatAlign
FlyMousePBMHuman <- getMatAlignOut("pcmsDatasetFlyMousePBMHuman",
                                   groupDistance=gpDis)
attach(FlyMousePBMHuman)

##set methods color
species <- factor(gsub("^(Dm|Mm|Hs).*$", "\\1", leaves))
##c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", "Mm/Uniprobe/PBM/Seed-And-Wobble")
levels(species) <- colorSet[c("Dm", "Hs", "Mm")] 
species <- as.character(species)

leaveNames <- gsub("^(Dm|Mm|Hs)_", "", leaves)

plotMotifStackWithRadialPhylog(phylog=phylog, pfms=sig, 
                               col.bg=species, col.bg.alpha=.3, 
                               col.inner.label.circle=gpCol, 
                               inner.label.circle.width=0.02, 
                               labels.leaves=leaveNames, 
                               cleaves=.1, circle=1.5, circle.motif=1.8, 
                               clabel.leaves=.5, motifScale="logarithmic", 
                               angle=358, plotIndex=TRUE, IndexCex=.6, 
                               groupDistance=gpDis,
                               groupDistanceLineCol="#D55E00")

legend(-2, 2.2, 
       legend=c("Dm/FFS/B1H", 
                "Mm/Uniprobe/PBM/Seed-And-Wobble", 
                "Hs/CIS-BP/SELEX"), 
       fill= highlightCol(colorSet[c("Dm", "Hs", "Mm")], alpha=.3), 
       border="white", lty=NULL, bty = "n")

detach(FlyMousePBMHuman)
```

```{r FlyMousePBMHuman.MotIV, fig.width=12, fig.height=12, dpi=72, fig.cap=fn("Alignment of motifs from fly, mouse and human by MotIV. Joint comparison of motifs with MotIV using a distance threshold of 0.005 with the same data sets and display settings as Supplemental Figure 3B.", "C"), echo=FALSE}
gpDis <- 0.005

# Supplementary Figure 3J. Align motifs from fly vs mouse and human by MotIV
FlyMousePBMHuman.MotIV <- getMotIVOut("pcmsDatasetFlyMousePBMHuman", 
                                      cc="ALLR", align="SWU", 
                                      groupDistance=gpDis)
attach(FlyMousePBMHuman.MotIV)

##set methods color
species <- factor(gsub("^(Dm|Mm|Hs).*$", "\\1", leaves))
##c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", "Mm/Uniprobe/PBM/Seed-And-Wobble")
levels(species) <- colorSet[c("Dm", "Hs", "Mm")] 
species <- as.character(species)

leaveNames <- gsub("^(Dm|Mm|Hs)_", "", leaves)

plotMotifStackWithRadialPhylog(phylog=phylog, pfms=sig, 
                               col.bg=species, col.bg.alpha=.3, 
                               col.inner.label.circle=gpCol, 
                               inner.label.circle.width=0.02, 
                               labels.leaves=leaveNames, 
                               cleaves=.1, circle=1.5, circle.motif=1.8, 
                               clabel.leaves=.5, motifScale="logarithmic", 
                               angle=358, plotIndex=TRUE, IndexCex=.6, 
                               groupDistance=gpDis)

legend(-2, 2.2, 
       legend=c("Dm/FFS/B1H", 
                "Mm/Uniprobe/PBM/Seed-And-Wobble", 
                "Hs/CIS-BP/SELEX"), 
       fill= highlightCol(colorSet[c("Dm", "Hs", "Mm")], alpha=.3), 
       border="white", lty=NULL, bty = "n")

detach(FlyMousePBMHuman.MotIV)
```

```{r EN, fig.width=8, fig.height=4, dpi=72, fig.cap=fn("HD motifs of Engrailed homologs from Drosophila, mouse and human illustrates how different experimental and computational methods can influence the resulting binding site motifs for orthologous proteins. Motif clusters are determined using a distance threshold of 2 (dashed magenta line). Dm/FFS/B1H (colored in turquoise) represents fly B1H binding site data. Hs/CIS-BP/SELEX (colored in purple) represents human SELEX data. Mm/Uniprobe/PBM/Seed-And-Wobble (colored in magenta) represents mouse PBM data analyzed by the Seed and Wobble motif discovery tool. Mm/CIS-BP/PBM/Dream5 (colored in blue) is mouse PBM data analyzed by the Dream5 motif discovery tool.", "D"), echo=FALSE, eval=TRUE}
gpDis <- 2

# Supplementary Figure 3D. Align motifs of engrailed orthologs by MatAlign
EN <- getMatAlignOut("pcmsDatasetEn", groupDistance=gpDis)
attach(EN)

##set methods color
species <- gsub("^(Dm|Hs|Mm).*?$", "\\1", leaves)
species[grepl("_M\\d+", leaves) & species=="Mm"] <- "MmDREAM"
species <- colorSet[species] ##c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", "Mm/Uniprobe/PBM/\nSeed-And-Wobble")

leaveNames <- gsub("^(Dm|Mm|Hs)_", "", leaves)

motifPiles(phylog=phylog, motifs, r.tree=.5, sig, 
           col.tree=species, col.leaves=species, 
           col.pfms2=gpCol, col.pfms2.width=.01, 
           labels.leaves=leaveNames, plotIndex=TRUE, 
           IndexCex=2, groupDistance=gpDis, clabel.leaves=2,
           groupDistanceLineCol="#D55E00")
legend("topleft", legend=c("Dm/FFS/B1H", "Hs/CIS-BP/SELEX", 
                           "Mm/Uniprobe/PBM/\nSeed-And-Wobble", 
                           "Mm/CIS-BP/PBM/Dream5"), 
       fill=colorSet[c("Dm", "Hs", "Mm", "MmDREAM")], 
       border="white", lty=NULL, bty = "n", cex=.5)

detach(EN)
```
